drivers/net/wireless/marvell/mwifiex/pcie_quirks.c

```diff
@@ -0,0 +1,248 @@
+	},
+	{
+		.ident = "Surface Pro 3",
+		.matches = {
+			DMI_EXACT_MATCH(DMI_SYS_VENDOR, "Microsoft Corporation"),
+			DMI_EXACT_MATCH(DMI_PRODUCT_NAME, "Surface Pro 3"),
+		},
+		.driver_data = (void *)QUIRK_FW_RST_WSID_SP3,
+	},
+	{}
+};
+
+void mwifiex_initialize_quirks(struct pcie_service_card *card)
+{
+	const struct dmi_system_id *dmi_id;
+
+	dmi_id = dmi_first_match(mwifiex_quirk_table);
+	if (dmi_id)
+		card->quirks = (uintptr_t)dmi_id->driver_data;
+
+	if (card->quirks & QUIRK_FW_RST_D3COLD)
+		mwifiex_dbg(card->adapter, MSG,
+			    "quirk reset_d3cold enabled\n");
+	if (card->quirks & QUIRK_FW_RST_WSID_S3)
+		mwifiex_dbg(card->adapter, MSG,
+			    "quirk reset_wsid for Surface 3 enabled\n");
+	if (card->quirks & QUIRK_FW_RST_WSID_SP3)
+		mwifiex_dbg(card->adapter, MSG,
+			    "quirk reset_wsid for Surface Pro 3 enabled\n");
+}
+
+static void mwifiex_pcie_set_power_d3cold(struct pci_dev *pdev)
+{
+	struct pcie_service_card *card = pci_get_drvdata(pdev);
+	struct mwifiex_adapter *adapter = card->adapter;
+
+	mwifiex_dbg(adapter, MSG, "putting into D3cold...\n");
+
+	pci_save_state(pdev);
+	if (pci_is_enabled(pdev))
+		pci_disable_device(pdev);
+	pci_set_power_state(pdev, PCI_D3cold);
+}
+
+static int mwifiex_pcie_set_power_d0(struct pci_dev *pdev)
+{
+	struct pcie_service_card *card = pci_get_drvdata(pdev);
+	struct mwifiex_adapter *adapter = card->adapter;
+	int ret;
+
+	mwifiex_dbg(adapter, MSG, "putting into D0...\n");
+
+	pci_set_power_state(pdev, PCI_D0);
+	ret = pci_enable_device(pdev);
+	if (ret) {
+		dev_err(&pdev->dev, "pci_enable_device failed\n");
+		return ret;
+	}
+	pci_restore_state(pdev);
+
+	return 0;
+}
+
+int mwifiex_pcie_reset_d3cold_quirk(struct pci_dev *pdev)
+{
+	struct pcie_service_card *card = pci_get_drvdata(pdev);
+	struct mwifiex_adapter *adapter = card->adapter;
+	struct pci_dev *parent_pdev = pci_upstream_bridge(pdev);
+
+	/* Power-cycle (put into D3cold then D0) */
+	mwifiex_dbg(adapter, MSG,
+		    "Using reset_d3cold quirk to perform FW reset\n");
+
+	/* We need to perform power-cycle also for bridge of wifi because
+	 * on some devices (e.g. Surface Book 1), the OS for some reasons
+	 * can't know the real power state of the bridge.
+	 * When tried to power-cycle only wifi, the reset failed with the
+	 * following dmesg log:
+	 * "Cannot transition to power state D0 for parent in D3hot".
+	 */
+	mwifiex_pcie_set_power_d3cold(pdev);
+	mwifiex_pcie_set_power_d3cold(parent_pdev);
+
+	mwifiex_pcie_set_power_d0(parent_pdev);
+	mwifiex_pcie_set_power_d0(pdev);
+
+	return 0;
+}
+
+int mwifiex_pcie_reset_wsid_quirk(struct pci_dev *pdev)
+{
+	struct pcie_service_card *card = pci_get_drvdata(pdev);
+	struct mwifiex_adapter *adapter = card->adapter;
+	char *path = ACPI_WSID_PATH;
+	static guid_t wsid_dsm_guid;
+	acpi_handle handle;
+	union acpi_object *obj;
+	acpi_status status;
+
+	mwifiex_dbg(adapter, MSG,
+		    "Using reset_wsid quirk to perform FW reset\n");
+
+	guid_parse(WSID_DSM_UUID, &wsid_dsm_guid);
```

Comment:
I think the preferred way is to use `GUID_INIT()` instead of defining the GUID as string, but this will work too. Not sure if upstream will complain about that.

Reply:
Switched to using GUID_INIT() by the new rebase.